### PR TITLE
Bug 1955517: Add cleanUpDuplicatedMC

### PR DESCRIPTION
**Pull request by QiWang19**
Fix BZ 1955517: upgrade when there are more than one configs,
these generated MC will be duplicated by upgraded MC with
number suffixed name (func getManagedKubeletConfigKey()) and fails the upgrade.
Remove the mc if its generated and degraded.

Signed-off-by: Qi Wang <qiwan@redhat.com>

**- What I did**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: pkg/controller/kubelet-config/helpers.go
Comment (umohnani8) [inline pkg/controller/kubelet-config/helpers.go:162]: We should add a condition for val so `if !ok || val == ""`. We will need to do the same for the container-runtime-config controller as well.

**Comment by QiWang19** [comment]
@umohnani8 PTAL. I tested the upgrade scenario as the Bug 1955517 mentioned. It can successfully upgrade from 4.6 to 4.7.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go]: fixed

**Comment by QiWang19** [comment]
github.com/pkg/errors

**Comment by QiWang19** [comment]
@kikisdeliveryservice @umohnani8 PTAL

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (haircommander) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:641]: I feel like I'm missing some information, but here are a couple of notes: 
- above we're skipping if it doesn't have the suffix, but the comment here says we're deleting if it doesn't have the suffix suffixed. 
- also, the comment says it will delete if degraded, but I'm not seeing us check if it is degraded.

**Comment by rphillips** [comment]
A unit test would be great to make sure the cleanup function works.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:641]: I should make the commet more clear, we will delete the MC without number suffix, those MC usually has ctrcfgMCSuffix.
The version.Hash is check the degradation.

Comment (umohnani8) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:641]: @QiWang19 I don't think we should just restrict this to non-suffixed MCs. We should delete any MC that is still using the old ignition version regardless of if it has a suffix or not.
I would suggest renaming `ctrcfgMCSuffix` to something like `generatedCtrcfg`  and check if the MC has that in its name.

Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go]: fixed

**Comment by QiWang19** [comment]
Added unit test for `cleanUpDuplicatedMC`.

**Comment by QiWang19** [comment]
@kikisdeliveryservice @umohnani8 @rphillips PTAL.

File: pkg/controller/kubelet-config/kubelet_config_controller.go
Comment (rphillips) [inline pkg/controller/kubelet-config/kubelet_config_controller.go:647]: does this cleanup a duplicated kubelet config? The function should be called cleanupDuplicatedKC

Comment (QiWang19) [reply pkg/controller/kubelet-config/kubelet_config_controller.go:647]: Only the machine config is duplicated, it only cleans up the machine config. The result of `oc get kubeletconfigs.machineconfiguration.openshift.io` doesn't change.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (umohnani8) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:642]: nit: can do `if mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey] != version.Hash` instead

File: pkg/controller/kubelet-config/kubelet_config_controller.go
Comment (umohnani8) [reply pkg/controller/kubelet-config/kubelet_config_controller.go:647]: You can create a common function for this in the common/helpers.go file as this is cleaning up MCs and is not specific to ctrcfg or kubeletconfig objects.

Comment (QiWang19) [reply pkg/controller/kubelet-config/kubelet_config_controller.go:647]: > You can create a common function for this in the common/helpers.go file as this is cleaning up MCs and is not specific to ctrcfg or kubeletconfig objects.

Didn't add it as a common, the cleanup functions are specific to different Controller.

File: pkg/controller/kubelet-config/kubelet_config_controller.go
Comment (umohnani8) [inline pkg/controller/kubelet-config/kubelet_config_controller.go:645]: nit: getManagedKeyCtrCfg() should be getManagedKubeletConfigKey()

File: pkg/controller/container-runtime-config/container_runtime_config_controller_test.go
Comment (umohnani8) [inline pkg/controller/container-runtime-config/container_runtime_config_controller_test.go:963]: @QiWang19 can you please add similar tests for kubeletconfig

**Comment by umohnani8** [comment]
LGTM after adding tests for kubeletconfig as well.

File: pkg/controller/container-runtime-config/container_runtime_config_controller_test.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller_test.go:963]: test added.

**Comment by QiWang19** [comment]
@kikisdeliveryservice @umohnani8 @rphillips PTAL.

**Comment by QiWang19** [comment]
Need approval label.

**Comment by kikisdeliveryservice** [comment]
@rphillips PTAL

Just holding my approval until Ryan gets a chance to approve. :)

**Comment by rphillips** [comment]
looks ready!

/lgtm

**Review by kikisdeliveryservice — APPROVED** [review]
let's get this in :)

**Pull request merged**
